### Pandas Lab -- Cleaning, Merging, & Grouping

This lab is designed to introduce students to common use cases for Pandas when working with data:

 - Creating new information out of your existing data set
 - Merging, concatenating, and joining different data sources
 - Grouping -- With both time & non-time based data

In [3]:
import pandas as pd
import numpy as np

In [4]:
df = pd.read_csv(r"/Users/PRSmb/OneDrive/General-Assembly/my-1019-repo/ClassMaterial/Unit2/data/restaurants.csv")

In [6]:
df.head()

id  visit_date  visitors calendar_date day_of_week  \
0  air_ba937bf13d40fb24  2016-01-13        25    2016-01-13   Wednesday   
1  air_ba937bf13d40fb24  2016-01-14        32    2016-01-14    Thursday   
2  air_ba937bf13d40fb24  2016-01-15        29    2016-01-15      Friday   
3  air_ba937bf13d40fb24  2016-01-16        22    2016-01-16    Saturday   
4  air_ba937bf13d40fb24  2016-01-18         6    2016-01-18      Monday   

   holiday       genre                          area   latitude   longitude  \
0        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
1        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
2        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
3        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
4        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   

   reserve_visitors  
0               NaN  
1               NaN  
2               NaN  
3               NaN  
4               NaN

### Section I: Creating Data Out of Your Existing Columns

Go ahead and create the following columns in your dataset.

**Column 1:**

  - **Column Name:** Weekend
  - **Values:** `True` if `day_of_week` is either Friday or Saturday, `False` if not

In [9]:
# your answer here

conditions = [
    df['day_of_week'] == 'Friday',
    df['day_of_week'] == "Saturday"
]

results = [
    True,
    True
]



df['Weekend'] = np.select(conditions, results, False)

In [10]:
df.head()

id  visit_date  visitors calendar_date day_of_week  \
0  air_ba937bf13d40fb24  2016-01-13        25    2016-01-13   Wednesday   
1  air_ba937bf13d40fb24  2016-01-14        32    2016-01-14    Thursday   
2  air_ba937bf13d40fb24  2016-01-15        29    2016-01-15      Friday   
3  air_ba937bf13d40fb24  2016-01-16        22    2016-01-16    Saturday   
4  air_ba937bf13d40fb24  2016-01-18         6    2016-01-18      Monday   

   holiday       genre                          area   latitude   longitude  \
0        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
1        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
2        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
3        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
4        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   

   reserve_visitors  Weekend  
0               NaN    False  
1               NaN    False  
2               NaN     True  
3               NaN     True  
4               NaN    False

**Column 2:**

 - **Column Name:** Reservation Activity
 - **Values:**
   - `Low` if `reserve_visitors` is in the bottom .25 percentile
   - `Medium` if `reserve_visitors` is in the middle .50 percentile
   - `High`if `reserve_visitors` is in the top .25 percentile
   
**Hint:** Use the `quantile` method to get this value

In [ ]:
# your answer here

In [27]:
df.reserve_visitors.quantile

<bound method Series.quantile of 0          NaN
1          NaN
2          NaN
3          NaN
4          NaN
          ... 
252103     6.0
252104    37.0
252105    35.0
252106     3.0
252107    32.0
Name: reserve_visitors, Length: 252108, dtype: float64>

In [31]:
?df['reserve_visitors'].quantile(q=[.25,.75])

Object `df['reserve_visitors'].quantile(q=[.25,.75])` not found.


In [23]:
conditions = [
    
]

results = [
    'Low',
    'Medium',
    'High'
]



df['reservation_activity'] = np.select(conditions, results, 'Unknown')

TypeError: float() argument must be a string or a number, not 'method'

**Column 3:**

 - **Column Name:** Days
 - **Values:**
   - The length of time that has passed from the beginning of the time series, in days
 - **Note:** When you subtract these columns, your column will be a **time delta**.  See if you can use the `dt` attribute to convert these values into an integer.  Ie, if your value reads `3 days`, you want that to be 3 instead.  You can read more about different time periods in pandas here:  https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#time-date-components

In [ ]:
# your answer here

### Section II: Merging Dataframes

The dataset we have been working with so far (`master.csv`), is actually a combined version of several datasets.  

In this section of the lab, we are going to re-create it manually from its individual pieces.

In the `restaurant data` folder, you'll find the following files:

 - `air_reserve.csv`
 - `air_store_info.csv`
 - `air_visit_data.csv`
 - `date_info.csv`
 
They contain all the constituent info for the `master.csv` file that we're currently using. 

You should have 252108 rows when you are finished.

Using merges, piece the files together to recreate the one we are currently working on.  

**Hint:** To get the number of reservations in the `reserve_visitors` column, you will have to use the `groupby` method first for each store_id and day before doing the merging.

You will also have to make sure each column is the same datatype.

Some operations that might come in handy:

 - `dt.date` -- converts a datetime to a date
 - `pd.to_datetime` if you need to convert something from a string to a date

In [ ]:
# your answer here